In [1]:
import pandas as pd

In [2]:
data = pd.read_excel("/content/Consulta_20240430-003636239.xlsx")

In [3]:
data

Fecha TIIE a 28 días, Tasa de interés en por ciento anual
0    2008-01-02                                               7.93 
1    2008-01-03                                             7.9362 
2    2008-01-04                                             7.9312 
3    2008-01-07                                             7.9275 
4    2008-01-08                                             7.9262 
...         ...                                                ... 
5861 2024-04-26                                             11.245 
5862 2024-04-27                                                N/E 
5863 2024-04-28                                                N/E 
5864 2024-04-29                                             11.245 
5865 2024-04-30                                             11.245 

[5866 rows x 2 columns]

In [6]:
datos = data

In [7]:
# Asegúrate de que los datos en la columna de interés sean numéricos, convirtiendo los no numéricos a NaN
datos['TIIE a 28 días, Tasa de interés en por ciento anual'] = pd.to_numeric(datos['TIIE a 28 días, Tasa de interés en por ciento anual'], errors='coerce')

# Calcula la media de los datos válidos (ignorando los NaN)
media_tiie = datos['TIIE a 28 días, Tasa de interés en por ciento anual'].mean()

# Reemplaza los valores NaN en la columna con la media calculada
datos['TIIE a 28 días, Tasa de interés en por ciento anual'].fillna(value=media_tiie, inplace=True)

datos

Fecha  TIIE a 28 días, Tasa de interés en por ciento anual
0    2008-01-02                                           7.930000  
1    2008-01-03                                           7.936200  
2    2008-01-04                                           7.931200  
3    2008-01-07                                           7.927500  
4    2008-01-08                                           7.926200  
...         ...                                                ...  
5861 2024-04-26                                          11.245000  
5862 2024-04-27                                           6.186164  
5863 2024-04-28                                           6.186164  
5864 2024-04-29                                          11.245000  
5865 2024-04-30                                          11.245000  

[5866 rows x 2 columns]

In [8]:
datos.to_excel("tasa_interes.xlsx")

In [9]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5866 entries, 0 to 5865
Data columns (total 2 columns):
 #   Column                                               Non-Null Count  Dtype         
---  ------                                               --------------  -----         
 0   Fecha                                                5866 non-null   datetime64[ns]
 1   TIIE a 28 días, Tasa de interés en por ciento anual  5866 non-null   float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 91.8 KB


In [19]:
datos


TIIE a 28 días, Tasa de interés en por ciento anual
Fecha                                                          
2008-01-02                                           7.930000  
2008-01-03                                           7.936200  
2008-01-04                                           7.931200  
2008-01-07                                           7.927500  
2008-01-08                                           7.926200  
...                                                       ...  
2024-04-26                                          11.245000  
2024-04-27                                           6.186164  
2024-04-28                                           6.186164  
2024-04-29                                          11.245000  
2024-04-30                                          11.245000  

[5866 rows x 1 columns]

In [20]:
datos.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 5866 entries, 2008-01-02 to 2024-04-30
Data columns (total 1 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   TIIE a 28 días, Tasa de interés en por ciento anual  5866 non-null   float64
dtypes: float64(1)
memory usage: 91.7 KB


In [21]:
# Resample los datos a nivel mensual y calcula el promedio
datos_mensuales = datos.resample('M').mean()

# Muestra los primeros resultados
print(datos_mensuales.head())

            TIIE a 28 días, Tasa de interés en por ciento anual
Fecha                                                          
2008-01-31                                           7.925986  
2008-02-29                                           7.928320  
2008-03-31                                           7.931206  
2008-04-30                                           7.935523  
2008-05-31                                           7.932700  


In [22]:
datos_mensuales

TIIE a 28 días, Tasa de interés en por ciento anual
Fecha                                                          
2008-01-31                                           7.925986  
2008-02-29                                           7.928320  
2008-03-31                                           7.931206  
2008-04-30                                           7.935523  
2008-05-31                                           7.932700  
...                                                       ...  
2023-12-31                                           9.445602  
2024-01-31                                           9.957893  
2024-02-29                                           9.847944  
2024-03-31                                           9.237462  
2024-04-30                                           9.897224  

[196 rows x 1 columns]

In [23]:
datos_mensuales.to_excel("timensual.xlsx")

In [32]:
data = pd.read_excel("/content/data diana.xlsx")
data

Fecha  tasa de interes  inflacion
0   2008-01-31         7.925986       3.70
1   2008-02-29         7.928320       3.72
2   2008-03-31         7.931206       4.25
3   2008-04-30         7.935523       4.55
4   2008-05-31         7.932700       4.95
..         ...              ...        ...
190 2023-11-30         9.732171       4.32
191 2023-12-31         9.445602       4.66
192 2024-01-31         9.957893       4.88
193 2024-02-29         9.847944       4.40
194 2024-03-31         9.237462       4.42

[195 rows x 3 columns]

In [33]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Fecha            195 non-null    datetime64[ns]
 1   tasa de interes  195 non-null    float64       
 2   inflacion        195 non-null    float64       
dtypes: datetime64[ns](1), float64(2)
memory usage: 4.7 KB


In [34]:
import pandas as pd
import numpy as np
from statsmodels.tsa.statespace.sarimax import SARIMAX
from statsmodels.tsa.stattools import adfuller

# Cargar datos
data = pd.read_excel('/content/data diana.xlsx')
data['Fecha'] = pd.to_datetime(data['Fecha'])
data.set_index('Fecha', inplace=True)

# Prueba de Dickey-Fuller para verificar estacionariedad
result = adfuller(data['tasa de interes'])
print('ADF Statistic: %f' % result[0])
print('p-value: %f' % result[1])

# Si no es estacionaria, diferenciar la serie
if result[1] > 0.05:
    data['Tasa_diff'] = data['tasa de interes'].diff().dropna()

# Identificar parámetros p, d, q usando ACF y PACF plots
# Aquí puedes añadir código para visualizar ACF y PACF si es necesario

# Ajustar el modelo ARIMAX
model = SARIMAX(data['Tasa_diff'],
                order=(1,1,1),  # estos son parámetros ejemplo
                exog=data['inflacion'],
                seasonal_order=(0,0,0,0),
                enforce_stationarity=False,
                enforce_invertibility=False)
results = model.fit()

# Ver el resumen del modelo
print(results.summary())


ADF Statistic: -2.299181
p-value: 0.172232
                               SARIMAX Results                                
Dep. Variable:              Tasa_diff   No. Observations:                  195
Model:               SARIMAX(1, 1, 1)   Log Likelihood                  53.872
Date:                Tue, 30 Apr 2024   AIC                            -99.743
Time:                        17:53:12   BIC                            -86.713
Sample:                    01-31-2008   HQIC                           -94.466
                         - 03-31-2024                                         
Covariance Type:                  opg                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
inflacion      0.0204      0.016      1.241      0.214      -0.012       0.053
ar.L1         -0.1819      0.077     -2.367      0.018      -0.332      -0.031
ma.L1    

/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency M will be used.
  self._init_dates(dates, freq)
/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency M will be used.
  self._init_dates(dates, freq)


Los resultados que has proporcionado incluyen la salida de una prueba de Dickey-Fuller aumentada (ADF) y un análisis del modelo SARIMAX (1, 1, 1) aplicado a una serie temporal de tasas de interés diferenciadas. Vamos a desglosar y entender cada parte de estos resultados:

### Prueba de Dickey-Fuller Aumentada (ADF)
- **Estadístico ADF**: -2.299181
- **Valor p**: 0.172232

La prueba ADF se utiliza para determinar si una serie temporal es estacionaria. El estadístico ADF de -2.299181 y un valor p de 0.172232 sugieren que **no podemos rechazar la hipótesis nula de que la serie tiene una raíz unitaria y, por lo tanto, no es estacionaria**. Esto implica que la serie temporal puede tener una tendencia o dependencia temporal.

### Modelo SARIMAX (1, 1, 1)
Este modelo intenta ajustar un modelo estacional autorregresivo integrado de media móvil con variables exógenas a la tasa de interés diferenciada.

- **Coeficientes**:
  - **Inflación**: El coeficiente para la inflación es 0.0204 con un error estándar de 0.016. Aunque positivo, el valor p de 0.214 indica que **no es estadísticamente significativo**.
  - **ar.L1**: El coeficiente autorregresivo de -0.1819 sugiere una **ligera autocorrelación negativa** en la serie y es estadísticamente significativo (p=0.018).
  - **ma.L1**: El coeficiente de media móvil de -0.6897 es significativo (p<0.001) e indica que el modelo ajusta considerablemente los errores del período anterior para predecir el valor actual.

- **Medidas de Ajuste**:
  - **Log Likelihood**: El logaritmo de la verosimilitud es de 53.872, lo que nos ayuda a entender la bondad del ajuste del modelo.
  - **AIC** y **BIC**: El AIC de -99.743 y el BIC de -86.713 son indicativos de que el modelo tiene un buen ajuste relativo, balanceando la complejidad y la bondad del ajuste.

### Diagnósticos del Modelo
- **Ljung-Box (Q)**: Un valor p de 0.84 sugiere que no hay autocorrelación significativa en los residuos, lo que es deseable en un buen modelo.
- **Jarque-Bera (JB)**: Un valor p de 0.00 indica que los residuos no siguen una distribución normal, lo que podría ser una preocupación. La prueba también indica alta kurtosis (5.43) y una ligera asimetría negativa.
- **Heteroskedasticidad (H)**: El test sugiere una ligera evidencia de heteroskedasticidad (p=0.14), lo que significa que la varianza de los residuos podría no ser constante.

### Conclusión
El modelo parece adecuado en términos de predecir la tasa de interés diferenciada basada en los datos anteriores y la inflación, aunque esta última no tiene un impacto significativo. Los problemas potenciales incluyen la no estacionariedad inicial de la serie, la no normalidad de los residuos y una posible heteroskedasticidad. Podría ser útil explorar modelos alternativos o realizar transformaciones adicionales en los datos para abordar estas preocupaciones.

**ARIMAX**

In [35]:
import pandas as pd
from statsmodels.tsa.statespace.sarimax import SARIMAX

# Supongamos que ya has cargado tus datos en un DataFrame llamado 'data'
# y que has establecido 'Fecha' como el índice del DataFrame

# Ejemplo de ajuste de un modelo ARIMAX
# Definimos los parámetros del modelo (estos deben ser seleccionados adecuadamente mediante análisis estadístico)
p = 1  # Orden de la parte autorregresiva
d = 1  # Orden de diferenciación
q = 1  # Orden de la parte de media móvil

# Ajustamos el modelo ARIMAX con la tasa de interés como variable dependiente y el índice de precios como exógena
model = SARIMAX(data['tasa de interes'],
                order=(p, d, q),
                exog=data['inflacion'],
                enforce_stationarity=False,
                enforce_invertibility=False)

results = model.fit(disp=False)  # Ajustamos el modelo sin mostrar mensajes

# Mostramos el resumen del modelo
print(results.summary())

# Para realizar predicciones, podríamos hacer algo como esto:
# pred = results.get_prediction(start=pd.to_datetime('2023-01-01'), dynamic=False)
# pred_ci = pred.conf_int()
# pred.predicted_mean.plot(ax=ax, label='One-step ahead Forecast', alpha=.7, figsize=(14, 7))
# ax.fill_between(pred_ci.index,
#                 pred_ci.iloc[:, 0],
#                 pred_ci.iloc[:, 1], color='k', alpha=.2)

# ax.set_xlabel('Fecha')
# ax.set_ylabel('Tasa de Interés')
# plt.legend()

# plt.show()


/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency M will be used.
  self._init_dates(dates, freq)
/usr/local/lib/python3.10/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: No frequency information was provided, so inferred frequency M will be used.
  self._init_dates(dates, freq)


                               SARIMAX Results                                
Dep. Variable:        tasa de interes   No. Observations:                  195
Model:               SARIMAX(1, 1, 1)   Log Likelihood                  55.199
Date:                Tue, 30 Apr 2024   AIC                           -102.398
Time:                        17:55:56   BIC                            -89.368
Sample:                    01-31-2008   HQIC                           -97.121
                         - 03-31-2024                                         
Covariance Type:                  opg                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
inflacion     -0.0229      0.038     -0.597      0.550      -0.098       0.052
ar.L1          0.9108      0.039     23.097      0.000       0.834       0.988
ma.L1         -1.4284      0.128    -11.158      0.0


### Interpretación de los resultados principales:

1. **Modelo**: ARIMAX(1, 1, 1)
   - **AR(1)**: El término autorregresivo (`ar.L1`) de 0.9108 sugiere que la tasa de interés del periodo actual tiene una fuerte correlación positiva con la tasa de interés del periodo anterior. Este coeficiente es estadísticamente significativo (`P>|z|` cercano a 0).
   - **MA(1)**: El término de media móvil (`ma.L1`) de -1.4284 indica que hay una corrección que se hace en base a errores pasados, y este también es un coeficiente significativo.

2. **Variables Exógenas**:
   - **Inflación**: El coeficiente de -0.0229 para la inflación indica que un aumento en la inflación está ligeramente asociado con una disminución en la tasa de interés, aunque este resultado no es estadísticamente significativo (`P>|z|` de 0.550).

3. **Información de Ajuste**:
   - **Log Likelihood**: El logaritmo de la verosimilitud es de 55.199, que ayuda a comparar modelos; a mayor valor, mejor ajuste del modelo.
   - **AIC** y **BIC**: El AIC es -102.398 y el BIC es -89.368. Ambos criterios buscan equilibrar la bondad de ajuste del modelo con la complejidad del mismo; valores más bajos generalmente indican un mejor modelo.

4. **Diagnósticos**:
   - **Ljung-Box (Q)**: Una prueba de Ljung-Box con un valor `Prob(Q)` de 0.12 sugiere que no hay evidencia suficiente para rechazar la hipótesis nula de que los residuos son independientes, es decir, el modelo parece capturar bien la autocorrelación.
   - **Jarque-Bera (JB)**: Un `Prob(JB)` de 0.00 indica que los residuos no siguen una distribución normal. La alta kurtosis y el sesgo indicado sugieren que los residuos tienen colas pesadas y pueden no ser simétricos.
   - **Heteroskedasticidad (H)**: El valor `Prob(H)` de 0.01 sugiere que hay evidencia de heteroscedasticidad en los residuos, lo cual podría indicar que la varianza de los residuos no es constante.


In [38]:
pip install arch


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 983.4/983.4 kB 5.8 MB/s eta 0:00:00


In [39]:
import pandas as pd
from arch import arch_model

# Cargar datos
data = pd.read_excel('/content/data diana.xlsx')
data['Fecha'] = pd.to_datetime(data['Fecha'])
data.set_index('Fecha', inplace=True)

# Seleccionar la serie de la tasa de interés
returns = data['tasa de interes']

# Ajustar un modelo ARCH(1)
model_arch = arch_model(returns, vol='Arch', p=1, q=0)  # ARCH(1)
res_arch = model_arch.fit(update_freq=5)

print(res_arch.summary())


Iteration:      5,   Func. Count:     27,   Neg. LLF: 1280.9282323135355
Iteration:     10,   Func. Count:     54,   Neg. LLF: 1193.443881994307
Iteration:     15,   Func. Count:     81,   Neg. LLF: 2839.3781219239736
Iteration:     20,   Func. Count:    106,   Neg. LLF: 226.65185705521156
Iteration:     25,   Func. Count:    134,   Neg. LLF: 1321.885356081257
Optimization terminated successfully    (Exit mode 0)
            Current function value: 207.34244143969488
            Iterations: 33
            Function evaluations: 158
            Gradient evaluations: 29
                      Constant Mean - ARCH Model Results                      
Dep. Variable:        tasa de interes   R-squared:                       0.000
Mean Model:             Constant Mean   Adj. R-squared:                  0.000
Vol Model:                       ARCH   Log-Likelihood:               -207.342
Distribution:                  Normal   AIC:                           420.685
Method:            Maximum Like

El resultado que compartiste es de un modelo ARCH (Autoregressive Conditional Heteroskedasticity) aplicado a la serie temporal de la tasa de interés. Este tipo de modelo se usa comúnmente para modelar y pronosticar la volatilidad de series temporales financieras. Aquí te dejo una interpretación detallada de cada parte de la salida:

### Proceso de Optimización
- **Iteraciones y Función de Log-Likelihood (LLF)**:
  - Las iteraciones muestran el proceso de minimización del logaritmo negativo de la función de verosimilitud (Neg. LLF). Comenzando en 1280.928 y reduciéndose significativamente hasta 226.651 y finalmente a 207.342.
  - `Func. Count` indica el número de evaluaciones de la función realizadas, y `Gradient evaluations` el número de evaluaciones del gradiente, que ayuda en la optimización numérica.

- **Convergencia**:
  - La optimización termina exitosamente después de 33 iteraciones, lo cual indica que el algoritmo de optimización encontró un punto mínimo local de la función de verosimilitud.

### Resultados del Modelo ARCH
- **Medidas de Ajuste**:
  - **Log-Likelihood**: -207.342 indica la bondad del ajuste del modelo a los datos.
  - **AIC** (Criterio de Información Akaike): 420.685 y **BIC** (Criterio de Información Bayesiano): 430.504 son medidas que penalizan la complejidad del modelo y cuanto más bajos, mejor.

- **Modelo de Media**:
  - **mu (media de la serie)**: El coeficiente es 5.2501 con un error estándar muy bajo, indicando que la media estimada de la tasa de interés es muy precisa y estadísticamente significativa (p-value de 0.000).

- **Modelo de Volatilidad**:
  - **omega**: El coeficiente de 0.0192 con un p-value de 0.411 sugiere que no es significativo, lo que indica que el término constante de la varianza del modelo puede no ser importante para explicar la variabilidad en los datos.
  - **alpha[1]**: Con un coeficiente de 0.9638 y un p-value extremadamente pequeño, indica que los rendimientos pasados tienen un impacto significativo y muy alto en la varianza actual, sugiriendo que la volatilidad de la tasa de interés es altamente persistente.

### Conclusión
El modelo ARCH indica que la tasa de interés tiene una media constante significativa y muestra alta volatilidad persistente, donde las fluctuaciones pasadas influyen fuertemente en la varianza futura. Esto es típico en datos financieros donde eventos pasados y shocks pueden tener un impacto duradero en la volatilidad. La significancia del parámetro alpha[1] pero no de omega sugiere que la variabilidad de la serie temporal es impulsada principalmente por su propia volatilidad pasada en lugar de una constante en el tiempo.

In [40]:
# Ajustar un modelo GARCH(1,1)
model_garch = arch_model(returns, vol='Garch', p=1, q=1)  # GARCH(1,1)
res_garch = model_garch.fit(update_freq=5)

print(res_garch.summary())


Iteration:      5,   Func. Count:     33,   Neg. LLF: 5463.011297084316
Iteration:     10,   Func. Count:     63,   Neg. LLF: 354.2594250005236
Iteration:     15,   Func. Count:     93,   Neg. LLF: 857.177330363404
Iteration:     20,   Func. Count:    122,   Neg. LLF: 199.269960312653
Optimization terminated successfully    (Exit mode 0)
            Current function value: 199.15728766867608
            Iterations: 23
            Function evaluations: 137
            Gradient evaluations: 23
                     Constant Mean - GARCH Model Results                      
Dep. Variable:        tasa de interes   R-squared:                       0.000
Mean Model:             Constant Mean   Adj. R-squared:                  0.000
Vol Model:                      GARCH   Log-Likelihood:               -199.157
Distribution:                  Normal   AIC:                           406.315
Method:            Maximum Likelihood   BIC:                           419.407
                             

El resultado que proporcionaste es de un modelo GARCH (Generalized Autoregressive Conditional Heteroskedasticity), que es una extensión del modelo ARCH y se utiliza comúnmente para modelar la volatilidad de series temporales financieras, en este caso, aplicado a la tasa de interés.

### Proceso de Optimización
- **Iteraciones y Función de Log-Likelihood (LLF)**:
  - Las iteraciones muestran la evolución del proceso de minimización de la función de logaritmo negativo de verosimilitud (Neg. LLF), que comienza con un valor alto de 5463.011 y reduce significativamente hasta 199.157.
  - `Func. Count` indica el número de veces que se evaluó la función para buscar el mínimo, con `Gradient evaluations` indicando el número de evaluaciones del gradiente, facilitando la optimización.

- **Convergencia**:
  - El proceso termina exitosamente tras 23 iteraciones, lo que indica que el algoritmo encontró un mínimo local satisfactorio de la función de verosimilitud.

### Resultados del Modelo GARCH
- **Medidas de Ajuste**:
  - **Log-Likelihood**: -199.157, que es una medida de la bondad del ajuste del modelo.
  - **AIC** (Criterio de Información Akaike): 406.315 y **BIC** (Criterio de Información Bayesiano): 419.407, que son medidas de selección de modelo que penalizan la complejidad. Cuanto más bajos, mejor es el balance entre la complejidad del modelo y el ajuste a los datos.

- **Modelo de Media**:
  - **mu (media de la serie)**: El coeficiente es 5.2702 con un error estándar muy bajo, indicando que la media estimada de la tasa de interés es muy precisa y estadísticamente significativa (p-value de 0.000).

- **Modelo de Volatilidad**:
  - **omega**: El coeficiente de 0.0051 con un p-value de 0.029 indica que el término constante de la varianza del modelo es significativo y relevante para la serie.
  - **alpha[1]**: Coeficiente de 1.0000 con un p-value extremadamente pequeño, indicando una alta significancia. Este valor sugiere una respuesta volátil completa a choques pasados en la serie.
  - **beta[1]**: Coeficiente de 0.0000 con un p-value de 1.000, lo que sugiere que el efecto de la volatilidad pasada en la volatilidad futura es no significativo, un resultado poco común y que podría necesitar revisión o reajuste del modelo.

### Conclusión
El modelo GARCH sugiere que la tasa de interés tiene una media constante muy significativa y muestra una volatilidad que responde completamente a los choques pasados, aunque el término beta es inusualmente no significativo, lo que podría indicar problemas en la especificación del modelo o en la naturaleza de los datos. El modelo indica que la varianza de la tasa de interés está influida principalmente por sus propios términos de volatilidad pasada y por nuevos choques, pero no por la persistencia de volatilidades pasadas. Este resultado puede requerir una revisión más profunda para entender mejor la dinámica de la serie temporal.